c10/core/TensorTypeSet.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include <c10/core/TensorTypeId.h>
+#include <c10/util/llvmMathExtras.h>
+#include <c10/util/Exception.h>
+#include <ostream>
+
+namespace c10 {
+
+// A representation of a set of TensorTypeIds.  A tensor may have multiple
+// tensor type ids, e.g., a Variable tensor can also be a CPU tensor; the
+// TensorTypeSet specifies what type ids apply.  The internal representation is
+// as a 64-bit bit set (this means only 64 tensor type ids are supported).
+//
+// Note that TensorTypeIds are ordered; thus, we can ask questions like "what is
+// the highest priority TensorTypeId in the set"?  (The set itself is not
+// ordered; two sets with the same ids will always have the ids ordered in the
+// same way.)
+//
+// At the moment, there are no nontrivial uses of this set; tensors are always
+// singletons.  In the near future, this set will represent variable? + tensor
+// type id.  In the far future, it will be requires grad? + profiling? +
+// tracing? + lazy? + tensor type id.
+//
+// (The difference between variable and requires grad, is that
+// there are currently three states a tensor can be:
+//  1. Not a variable
+//  2. Variable with requires_grad=False
+//  3. Variable with requires_grad=True
+// Eventually, we want to kill state (1), and only dispatch to autograd
+// handling code if one of the inputs requires grad.)
+//
+// An undefined tensor is one with an empty tensor type set.
+class TensorTypeSet {
+public:
+  TensorTypeSet() {}
+  explicit TensorTypeSet(TensorTypeId t)
+    : repr_(t == TensorTypeId::UndefinedTensorId
+              ? 0
+              : 1ULL << (static_cast<uint8_t>(t) - 1)) {}
+  // Test if a TensorTypeId is in the set
+  bool has(TensorTypeId t) const {
+    TORCH_INTERNAL_ASSERT(t != TensorTypeId::UndefinedTensorId);
+    return static_cast<bool>(repr_ & TensorTypeSet(t).repr_);
+  }
+  // Perform set union
+  TensorTypeSet operator|(TensorTypeSet other) const {
```

Comment:
I think `operator|` is the wrong thing here, it exposes the implementation detail of it being a bitfield in the API. Semantically, these are sets. Let's call it `merge()` or `add()` or `append()`?

Reply:
I disagree. Python defines set union to be overloaded with `|`, so this usage is consistent.